ziro-export-risport.ps1

```diff
@@ -0,0 +1,230 @@
+   }
+}
+
+function Get-Devices {
+   param (
+      [CucmConnector]$CucmConnector
+   )
+
+   $allDevices = New-Object System.Collections.Generic.List[object]
+   $skip = 0
+   $first = 2000
+   $response = $CucmConnector.ListPhone($skip, $first)
+   
+   while (!($response.GetElementsByTagName('return').IsEmpty)) {
+      $pageDevices = [object[]]($response.GetElementsByTagName('return').phone)
+      $allDevices.AddRange($pageDevices)
+      $skip += $first
+      $response = $CucmConnector.ListPhone($skip, $first)
+   }
+
+   return $allDevices
+}
+
+function Get-DeviceRegistrationStatuses {
+   param (
+      [string[]]$DeviceNames,
+      [CucmConnector]$CucmConnector
+   )
+
+   $registrationStatuses = @{}
+   $maxReturnedDevices = 2000  # The endpoint allows querying a max of 2000 devices at a time
+   $windowStartIndex = 0
+
+   while ($windowStartIndex -lt $DeviceNames.Length) {
+      $progressParameters = @{
+         Activity        = 'Fetching device registrations statuses...'
+         Status          = "Fetched: $windowStartIndex of $($DeviceNames.Length)"
+         PercentComplete = ($windowStartIndex / $DeviceNames.Length) * 100
+      }
+
+      Write-Progress @progressParameters
+
+      $windowEndIndex = $windowStartIndex + $maxReturnedDevices - 1
+      # powershell doesn't throw an out-of-bounds error when accessing an index that is past the end of array
+      $devicesToQuery = $DeviceNames[$windowStartIndex..$windowEndIndex]  # upper bound is inclusive
+      $response = $CucmConnector.SelectCmDevice($devicesToQuery)
+      $totalDevicesFound = $response.Envelope.Body.selectCmDeviceResponse.selectCmDeviceReturn.SelectCmDeviceResult.TotalDevicesFound
+
+      if ($totalDevicesFound -gt 0) {
+         $devicesPerNode = $response.Envelope.Body.selectCmDeviceResponse.selectCmDeviceReturn.SelectCmDeviceResult.CmNodes.item.CmDevices 
+         | Where-Object { $_.HasChildNodes }
+
+         foreach ($nodeDevices in $devicesPerNode) {
+            $nodeDevices.item | Select-Object -Property Name, Status, StatusReason, TimeStamp 
+            | ForEach-Object { $registrationStatuses.Add($_.Name, $_) }
+         }
+      }
+
+      $windowStartIndex = $windowStartIndex + $maxReturnedDevices
+      Start-Sleep -Seconds 1  # a pinch of throttling to help with rate limits
+   }
+
+   return $registrationStatuses
+}
+
+
+$serverUrl = Read-Host 'CUCM server URL (https://mycucm.company.com)'
+$credential = Get-Credential -Message 'Enter username and password'
+$cucmConnector = [CucmConnector]::new($serverUrl, $credential) 
+
+try {
+   Write-Host "Fetching devices from CUCM server..."
+   $devices = Get-Devices -CucmConnector $cucmConnector
+   Write-Host "Found [$($devices.Length)] devices"
+   Write-Host "Getting device registration statuses..."
+   $registrationStatuses = Get-DeviceRegistrationStatuses -DeviceNames $devices.name -CucmConnector $cucmConnector
+   $registrationStatuses
+   $registrationStatuses
```

Comment:
what's the purpose of duplicating this line?

Reply:
idk wtf happened - deleted